crates/red_knot/src/types.rs

```diff
@@ -37,40 +38,137 @@ impl TypeEnvironment {
     }
 }
 
+#[derive(Debug)]
 pub(crate) enum Type {
-    Class(ClassType),
+    // a specific function
     Function(FunctionType),
+    // the set of Python objects with this class in their MRO
+    Instance(ClassType),
+    Union(UnionType),
+    Intersection(IntersectionType),
+    // the dynamic or gradual type: a statically-unknown set of values
+    Any,
```

Comment:
Any, Unknown and Unbound seem to all be types that represent dynamic or unsafe typing in some way. That reminds me of this in the mypy codebase (which is really an enum, though it doesn't use Python's `enum` module, I think for performance and/or mypyc reasons): https://github.com/python/mypy/blob/82ebd866830cd79e25bf9d59e9f9474bd280c4f5/mypy/types.py#L187

I wonder if it might make sense to have a single Dynamic variant here:

```suggestion
    Dynamic(DynamicType),
```

And have a separate DynamicType enum, e.g.

```rs
/// Enumeration of types that are fundamentally unsafe in some way
enum DynamicType {
    Any,
    Unbound,
    Unknown
}
```

Reply:
I think the semantics of `Unbound` are quite different; it doesn't represent dynamic typing, it represents cases where we know for sure that a name either is unbound, or may be unbound (in this case we'd use a union with unbound). So I wouldn't want to include `Unbound` in such a `DynamicType` enum.

(FWIW, I'm also not totally sold on representing `Unbound` as a type at all. It will result in a lot more union types, and maybe confusing error messages too. And I don't think it fully fits into the model, since a name can be unbound, but an expression can't be. So I'm leaning towards just removing `Unbound` here and modeling it differently.)

We could have a `DynamicType` enum, but it also doesn't seem clearly necessary, so I'd rather wait and see if the type evaluation code suggests that it would be beneficial in some way.

For instance, if we want to have a strict mode where `Unknown` acts like `object`, then I think grouping `Any` and `Unknown` together like this might cause more harm than good.

It may be that at some point we want to track more fine-grained versions of `Any` like mypy does; at that point I think grouping them would make more sense.